Failed to parse the media type in 'accept' header: *; q=.2

I use armeria-tomcat service, version 0.87, Configures an Armeria to redirect the incoming requests to the Tomcat instance provided by Spring Boot. like: examples\spring-boot-tomcat
And Client use java.net.HttpURLConnection, but found below failed issue:

019-06-18 16:14:06.728 [armeria-common-worker-nio-5-1] DEBUG com.linecorp.armeria.server.DefaultRoutingContext - Failed to parse the media type in 'accept' header: *; q=.2
java.lang.IllegalArgumentException: Could not parse '*; q=.2'
	at com.linecorp.armeria.common.MediaType.parse(MediaType.java:994)
	at com.linecorp.armeria.server.DefaultRoutingContext.lambda$extractAcceptTypes$0(DefaultRoutingContext.java:177)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:580)
	at com.linecorp.armeria.server.DefaultRoutingContext.lambda$extractAcceptTypes$1(DefaultRoutingContext.java:174)
	at com.linecorp.armeria.internal.shaded.guava.collect.ImmutableList.forEach(ImmutableList.java:405)
	at com.linecorp.armeria.server.DefaultRoutingContext.extractAcceptTypes(DefaultRoutingContext.java:173)
	at com.linecorp.armeria.server.DefaultRoutingContext.of(DefaultRoutingContext.java:57)
	at com.linecorp.armeria.server.HttpServerHandler.handleRequest(HttpServerHandler.java:329)
	at com.linecorp.armeria.server.HttpServerHandler.channelRead(HttpServerHandler.java:250)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at com.linecorp.armeria.server.Http1RequestDecoder.channelRead(Http1RequestDecoder.java:185)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:310)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:426)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.codec.ByteToMessageDecoder.handlerRemoved(ByteToMessageDecoder.java:249)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:507)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.flush.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:154)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:287)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:352)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:374)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:360)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:682)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:617)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:534)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:496)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:906)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: null
	at com.linecorp.armeria.internal.shaded.guava.base.Preconditions.checkState(Preconditions.java:492)
	at com.linecorp.armeria.common.MediaType$Tokenizer.consumeCharacter(MediaType.java:1030)
	at com.linecorp.armeria.common.MediaType.parse(MediaType.java:964)
	... 62 common frames omitted
2019-06-18 16:14:06.790 [armeria-common-worker-nio-5-1] INFO  armeria.services.api.test - [id: 0xf2709a41, L:/127.0.0.1:8096 - R:/127.0.0.1:64177][h1c://xxxxxx:8096/api/test/login#POST] Request: {startTime=2019-06-18T08:14:06.743Z(1560845646743000), length=114B, duration=24607µs(24607248ns), scheme=none+h1c, headers=[:method=POST, :path=/api/test/login, :scheme=http, :authority=localhost:8096, content-type=application/json, authorization=Bearer 123, user-agent=Java/1.8.0_161, accept=text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2, content-length=114]}

what can I do? thank.

--- a/core/src/main/java/com/linecorp/armeria/server/DefaultRoute.java
+++ b/core/src/main/java/com/linecorp/armeria/server/DefaultRoute.java
@@ -96,26 +96,30 @@ final class DefaultRoute implements Route {
             return RoutingResult.empty();
         }
 
-        if (!consumes.isEmpty()) {
-            final MediaType contentType = routingCtx.contentType();
-            boolean found = false;
-            if (contentType != null) {
-                for (MediaType consumeType : consumes) {
-                    found = contentType.belongsTo(consumeType);
-                    if (found) {
-                        break;
-                    }
+        final MediaType contentType = routingCtx.contentType();
+        boolean contentTypeMatched = false;
+        if (contentType == null) {
+            if (consumes().isEmpty()) {
+                contentTypeMatched = true;
+            }
+        } else if (!consumes().isEmpty()) {
+            for (MediaType consumeType : consumes) {
+                contentTypeMatched = contentType.belongsTo(consumeType);
+                if (contentTypeMatched) {
+                    break;
                 }
             }
-            if (!found) {
+            if (!contentTypeMatched) {
                 routingCtx.delayThrowable(HttpStatusException.of(HttpStatus.UNSUPPORTED_MEDIA_TYPE));
                 return RoutingResult.empty();
             }
         }
 
-        final List<MediaType> types = routingCtx.acceptTypes();
-        if (types.isEmpty()) {
-            builder.score(HIGHEST_SCORE);
+        final List<MediaType> acceptTypes = routingCtx.acceptTypes();
+        if (acceptTypes.isEmpty()) {
+            if (contentTypeMatched && produces().isEmpty()) {
+                builder.score(HIGHEST_SCORE);
+            }
             for (MediaType produceType : produces) {
                 if (!isAnyType(produceType)) {
                     return builder.negotiatedResponseMediaType(produceType).build();
@@ -126,8 +130,8 @@ final class DefaultRoute implements Route {
 
         if (!produces.isEmpty()) {
             for (MediaType produceType : produces) {
-                for (int i = 0; i < types.size(); i++) {
-                    final MediaType acceptType = types.get(i);
+                for (int i = 0; i < acceptTypes.size(); i++) {
+                    final MediaType acceptType = acceptTypes.get(i);
                     if (produceType.belongsTo(acceptType)) {
                         // To early stop path mapping traversal,
                         // we set the score as the best score when the index is 0.

--- a/core/src/main/java/com/linecorp/armeria/server/DefaultRoutingContext.java
+++ b/core/src/main/java/com/linecorp/armeria/server/DefaultRoutingContext.java
@@ -53,36 +53,29 @@ final class DefaultRoutingContext implements RoutingContext {
     static RoutingContext of(VirtualHost virtualHost, String hostname,
                              String path, @Nullable String query,
                              RequestHeaders headers, boolean isCorsPreflight) {
-        return new DefaultRoutingContext(virtualHost, hostname, headers.method(), path, query,
-                                         headers.contentType(), extractAcceptTypes(headers),
-                                         isCorsPreflight);
+        return new DefaultRoutingContext(virtualHost, hostname, headers, path, query, isCorsPreflight);
     }
 
     private final VirtualHost virtualHost;
     private final String hostname;
-    private final HttpMethod method;
+    private final RequestHeaders headers;
     private final String path;
     @Nullable
     private final String query;
     @Nullable
-    private final MediaType contentType;
-    private final List<MediaType> acceptTypes;
+    private volatile List<MediaType> acceptTypes;
     private final boolean isCorsPreflight;
     private final List<Object> summary;
     @Nullable
     private Throwable delayedCause;
 
-    DefaultRoutingContext(VirtualHost virtualHost, String hostname,
-                          HttpMethod method, String path, @Nullable String query,
-                          @Nullable MediaType contentType, List<MediaType> acceptTypes,
-                          boolean isCorsPreflight) {
+    DefaultRoutingContext(VirtualHost virtualHost, String hostname, RequestHeaders headers,
+                          String path, @Nullable String query, boolean isCorsPreflight) {
         this.virtualHost = requireNonNull(virtualHost, "virtualHost");
         this.hostname = requireNonNull(hostname, "hostname");
-        this.method = requireNonNull(method, "method");
+        this.headers = requireNonNull(headers, "headers");
         this.path = requireNonNull(path, "path");
         this.query = query;
-        this.contentType = contentType;
-        this.acceptTypes = requireNonNull(acceptTypes, "acceptTypes");
         this.isCorsPreflight = isCorsPreflight;
         summary = generateSummary(this);
     }
@@ -99,7 +92,7 @@ final class DefaultRoutingContext implements RoutingContext {
 
     @Override
     public HttpMethod method() {
-        return method;
+        return headers.method();
     }
 
     @Override
@@ -116,11 +109,16 @@ final class DefaultRoutingContext implements RoutingContext {
     @Nullable
     @Override
     public MediaType contentType() {
-        return contentType;
+        return headers.contentType();
     }
 
     @Override
     public List<MediaType> acceptTypes() {
+        List<MediaType> acceptTypes = this.acceptTypes;
+        if (acceptTypes == null) {
+            acceptTypes = extractAcceptTypes(headers);
+            this.acceptTypes = acceptTypes;
+        }
         return acceptTypes;
     }
 
@@ -176,8 +174,8 @@ final class DefaultRoutingContext implements RoutingContext {
                             try {
                                 acceptTypes.add(MediaType.parse(mediaType));
                             } catch (IllegalArgumentException e) {
-                                logger.debug("Failed to parse the media type in 'accept' header: {}",
-                                             mediaType, e);
+                                logger.debug("Ignoring a malformed media type from 'accept' header: {}",
+                                             mediaType);
                             }
                         }));
         if (acceptTypes.isEmpty()) {

--- a/core/src/main/java/com/linecorp/armeria/server/Routers.java
+++ b/core/src/main/java/com/linecorp/armeria/server/Routers.java
@@ -251,11 +251,11 @@ public final class Routers {
                     //
                     // The services are sorted as follows:
                     //
-                    // 1) annotated service with method and media type negotiation
+                    // 1) the service with method and media type negotiation
                     //    (consumable and producible)
-                    // 2) annotated service with method and producible media type negotiation
-                    // 3) annotated service with method and consumable media type negotiation
-                    // 4) annotated service with method negotiation
+                    // 2) the service with method and producible media type negotiation
+                    // 3) the service with method and consumable media type negotiation
+                    // 4) the service with method negotiation
                     // 5) the other services (in a registered order)
                     //
                     // 1) and 2) may produce a score between the lowest and the highest because they should
@@ -266,15 +266,6 @@ public final class Routers {
                     // Found the best matching.
                     if (routingResult.hasHighestScore()) {
                         result = Routed.of(route, routingResult, value);
-                        break;
-                    }
-
-                    // This means that the 'routingResult' is produced by one of 3), 4) and 5).
-                    // So we have no more chance to find a better matching from now.
-                    if (routingResult.hasLowestScore()) {
-                        if (!result.isPresent()) {
-                            result = Routed.of(route, routingResult, value);
-                        }
                         break;
                     }
 

--- a/core/src/main/java/com/linecorp/armeria/server/RoutingContext.java
+++ b/core/src/main/java/com/linecorp/armeria/server/RoutingContext.java
@@ -97,8 +97,12 @@ public interface RoutingContext {
      */
     default RoutingContext overridePath(String path) {
         requireNonNull(path, "path");
-        return new DefaultRoutingContext(virtualHost(), hostname(), method(), path, query(),
-                                         contentType(), acceptTypes(), isCorsPreflight());
+        return new RoutingContextWrapper(this) {
+            @Override
+            public String path() {
+                return path;
+            }
+        };
     }
 
     /**

--- a/core/src/main/java/com/linecorp/armeria/server/RoutingContextWrapper.java
+++ b/core/src/main/java/com/linecorp/armeria/server/RoutingContextWrapper.java
@@ -1,0 +1,91 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.MediaType;
+
+class RoutingContextWrapper implements RoutingContext {
+
+    private final RoutingContext delegate;
+
+    RoutingContextWrapper(RoutingContext delegate) {
+        this.delegate = delegate;
+    }
+
+    @Override
+    public VirtualHost virtualHost() {
+        return delegate.virtualHost();
+    }
+
+    @Override
+    public String hostname() {
+        return delegate.hostname();
+    }
+
+    @Override
+    public HttpMethod method() {
+        return delegate.method();
+    }
+
+    @Override
+    public String path() {
+        return delegate.path();
+    }
+
+    @Nullable
+    @Override
+    public String query() {
+        return delegate.query();
+    }
+
+    @Nullable
+    @Override
+    public MediaType contentType() {
+        return delegate.contentType();
+    }
+
+    @Override
+    public List<MediaType> acceptTypes() {
+        return delegate.acceptTypes();
+    }
+
+    @Override
+    public List<Object> summary() {
+        return delegate.summary();
+    }
+
+    @Override
+    public void delayThrowable(Throwable cause) {
+        delegate.delayThrowable(cause);
+    }
+
+    @Override
+    public Optional<Throwable> delayedThrowable() {
+        return delegate.delayedThrowable();
+    }
+
+    @Override
+    public boolean isCorsPreflight() {
+        return delegate.isCorsPreflight();
+    }
+}

--- a/core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/annotation/AnnotatedHttpServiceTest.java
@@ -787,12 +787,11 @@ public class AnnotatedHttpServiceTest {
         final HttpClient client = HttpClient.of(rule.uri("/"));
         final String path = "/8/same/path";
 
-        // No 'Accept' header means accepting everything. The order of -1 would be matched first.
         RequestHeaders headers = RequestHeaders.of(HttpMethod.GET, path);
         AggregatedHttpResponse res = client.execute(headers).aggregate().join();
         assertThat(res.status()).isSameAs(HttpStatus.OK);
-        assertThat(res.headers().contentType()).hasToString("text/plain");
-        assertThat(res.contentUtf8()).isEqualTo("GET/TEXT");
+        assertThat(res.headers().contentType()).isNull();
+        assertThat(res.contentUtf8()).isEqualTo("GET");
 
         // The same as the above.
         headers = RequestHeaders.of(HttpMethod.GET, path, HttpHeaderNames.ACCEPT, "*/*");
@@ -854,15 +853,15 @@ public class AnnotatedHttpServiceTest {
                                     HttpHeaderNames.CONTENT_TYPE, "application/json");
         res = client.execute(headers).aggregate().join();
         assertThat(res.status()).isSameAs(HttpStatus.OK);
-        assertThat(res.headers().contentType()).isSameAs(MediaType.JSON);
-        assertThat(res.contentUtf8()).isEqualTo("POST/JSON/BOTH");
+        assertThat(res.headers().contentType()).isNull();
+        assertThat(res.contentUtf8()).isEqualTo("POST/JSON");
 
         headers = RequestHeaders.of(HttpMethod.POST, path,
                                     HttpHeaderNames.ACCEPT, "application/json");
         res = client.execute(headers).aggregate().join();
         assertThat(res.status()).isSameAs(HttpStatus.OK);
-        assertThat(res.headers().contentType()).isNull();
-        assertThat(res.contentUtf8()).isEqualTo("POST");
+        assertThat(res.headers().contentType()).isSameAs(MediaType.JSON);
+        assertThat(res.contentUtf8()).isEqualTo("POST/JSON/BOTH");
 
         headers = RequestHeaders.of(HttpMethod.POST, path);
         res = client.execute(headers).aggregate().join();

--- a/core/src/test/java/com/linecorp/armeria/server/RouteTest.java
+++ b/core/src/test/java/com/linecorp/armeria/server/RouteTest.java
@@ -232,8 +232,6 @@ class RouteTest {
                      .build();
         getResult = route.apply(method(HttpMethod.GET));
         assertThat(getResult.isPresent()).isTrue();
-        // When there's no "Accept" header, it has the highest score.
-        assertThat(getResult.hasHighestScore()).isTrue();
 
         getResult = route.apply(withAcceptHeader(HttpMethod.GET, "application/json;charset=UTF-8"));
         assertThat(getResult.isPresent()).isFalse();
@@ -254,8 +252,6 @@ class RouteTest {
 
         getResult = route.apply(method(HttpMethod.GET));
         assertThat(getResult.isPresent()).isTrue();
-        // When there's no "Accept" header, it has the highest score.
-        assertThat(getResult.hasHighestScore()).isTrue();
 
         getResult = route.apply(withAcceptHeader(HttpMethod.GET, "*/*"));
         assertThat(getResult.isPresent()).isTrue();

--- a/core/src/test/java/com/linecorp/armeria/server/RoutingContextTest.java
+++ b/core/src/test/java/com/linecorp/armeria/server/RoutingContextTest.java
@@ -27,8 +27,6 @@ import javax.annotation.Nullable;
 
 import org.junit.jupiter.api.Test;
 
-import com.google.common.collect.ImmutableList;
-
 import com.linecorp.armeria.common.HttpHeaderNames;
 import com.linecorp.armeria.common.HttpHeaders;
 import com.linecorp.armeria.common.HttpMethod;
@@ -77,31 +75,36 @@ class RoutingContextTest {
         final RoutingContext ctx3;
 
         ctx1 = new DefaultRoutingContext(virtualHost, "example.com",
-                                         HttpMethod.GET, "/hello", null,
-                                         MediaType.JSON_UTF_8,
-                                         ImmutableList.of(MediaType.JSON_UTF_8, MediaType.XML_UTF_8),
-                                         false);
+                                         RequestHeaders.of(HttpMethod.GET, "/hello",
+                                                           HttpHeaderNames.CONTENT_TYPE, MediaType.JSON_UTF_8,
+                                                           HttpHeaderNames.ACCEPT,
+                                                           MediaType.JSON_UTF_8 + ", " +
+                                                           MediaType.XML_UTF_8 + "; q=0.8"),
+                                         "/hello", null, false);
         ctx2 = new DefaultRoutingContext(virtualHost, "example.com",
-                                         HttpMethod.GET, "/hello", null,
-                                         MediaType.JSON_UTF_8,
-                                         ImmutableList.of(MediaType.JSON_UTF_8, MediaType.XML_UTF_8),
-                                         false);
+                                         RequestHeaders.of(HttpMethod.GET, "/hello",
+                                                           HttpHeaderNames.CONTENT_TYPE, MediaType.JSON_UTF_8,
+                                                           HttpHeaderNames.ACCEPT,
+                                                           MediaType.JSON_UTF_8 + ", " +
+                                                           MediaType.XML_UTF_8 + "; q=0.8"),
+                                         "/hello", null, false);
         ctx3 = new DefaultRoutingContext(virtualHost, "example.com",
-                                         HttpMethod.GET, "/hello", null,
-                                         MediaType.JSON_UTF_8,
-                                         ImmutableList.of(MediaType.XML_UTF_8, MediaType.JSON_UTF_8),
-                                         false);
+                                         RequestHeaders.of(HttpMethod.GET, "/hello",
+                                                           HttpHeaderNames.CONTENT_TYPE, MediaType.JSON_UTF_8,
+                                                           HttpHeaderNames.ACCEPT,
+                                                           MediaType.XML_UTF_8 + ", " +
+                                                           MediaType.JSON_UTF_8 + "; q=0.8"),
+                                         "/hello", null, false);
 
         assertThat(ctx1.hashCode()).isEqualTo(ctx2.hashCode());
         assertThat(ctx1).isEqualTo(ctx2);
         assertThat(ctx1).isNotEqualTo(ctx3);
 
         ctx1 = new DefaultRoutingContext(virtualHost, "example.com",
-                                         HttpMethod.GET, "/hello", "a=1&b=1",
-                                         null, ImmutableList.of(), false);
+                                         RequestHeaders.of(HttpMethod.GET, "/hello"),
+                                         "/hello", "a=1&b=1", false);
         ctx2 = new DefaultRoutingContext(virtualHost, "example.com",
-                                         HttpMethod.GET, "/hello", "a=1",
-                                         null, ImmutableList.of(), false);
+                                         RequestHeaders.of(HttpMethod.GET, "/hello"), "/hello", "a=1", false);
 
         assertThat(ctx1.hashCode()).isEqualTo(ctx2.hashCode());
         assertThat(ctx1).isEqualTo(ctx2);

--- a/core/src/test/java/com/linecorp/armeria/server/ServiceBindingTest.java
+++ b/core/src/test/java/com/linecorp/armeria/server/ServiceBindingTest.java
@@ -105,11 +105,8 @@ class ServiceBindingTest {
                           final MediaType contentType = req.contentType();
                           if (contentType == MediaType.JSON) {
                               resContent = "{\"name\":\"" + request.contentUtf8() + "\"}";
-                          } else if (contentType == MediaType.PLAIN_TEXT_UTF_8) {
+                          } else  {
                               resContent = request.contentUtf8();
-                          } else {
-                              fail("Should never reach here");
-                              resContent = "Never reach here.";
                           }
                           return HttpResponse.of(resContent);
                       })));
@@ -140,7 +137,8 @@ class ServiceBindingTest {
         final HttpClient client = HttpClient.of(server.uri("/"));
         AggregatedHttpResponse res = client.execute(RequestHeaders.of(HttpMethod.POST, "/hello"), "armeria")
                                           .aggregate().join();
-        assertThat(res.status()).isSameAs(HttpStatus.UNSUPPORTED_MEDIA_TYPE);
+        assertThat(res.status()).isSameAs(HttpStatus.OK);
+        assertThat(res.contentUtf8()).isEqualTo("armeria");
 
         res = client.execute(RequestHeaders.of(HttpMethod.POST, "/hello",
                                                HttpHeaderNames.CONTENT_TYPE, MediaType.PLAIN_TEXT_UTF_8),

--- a/core/src/test/java/com/linecorp/armeria/server/cors/HttpServerCorsTest.java
+++ b/core/src/test/java/com/linecorp/armeria/server/cors/HttpServerCorsTest.java
@@ -43,6 +43,7 @@ import com.linecorp.armeria.server.ServerBuilder;
 import com.linecorp.armeria.server.Service;
 import com.linecorp.armeria.server.ServiceRequestContext;
 import com.linecorp.armeria.server.annotation.AdditionalHeader;
+import com.linecorp.armeria.server.annotation.ConsumesJson;
 import com.linecorp.armeria.server.annotation.Get;
 import com.linecorp.armeria.server.annotation.Options;
 import com.linecorp.armeria.server.annotation.Param;
@@ -56,19 +57,20 @@ public class HttpServerCorsTest {
 
     private static final ClientFactory clientFactory = ClientFactory.DEFAULT;
 
-    @CorsDecorators(value = {
-            @CorsDecorator(origins = "http://example.com", exposedHeaders = "expose_header_1")
-    }, shortCircuit = true)
+    @CorsDecorators(value = @CorsDecorator(origins = "http://example.com", exposedHeaders = "expose_header_1"),
+            shortCircuit = true)
     private static class MyAnnotatedService {
 
         @Get("/index")
         @StatusCode(200)
         public void index() {}
 
+        @ConsumesJson
         @Get("/dup_test")
         @StatusCode(200)
-        @CorsDecorator(origins = "http://example2.com", exposedHeaders = "expose_header_2")
-        public void duptest() {}
+        @CorsDecorator(origins = "http://example2.com", exposedHeaders = "expose_header_2",
+                allowedRequestHeaders = "content-type")
+        public void dupTest() {}
     }
 
     @CorsDecorator(
@@ -170,37 +172,30 @@ public class HttpServerCorsTest {
             sb.annotatedService("/cors6", new Object() {
 
                 @Get("/any/get")
-                @CorsDecorator(
-                        origins = "*", exposedHeaders = { "expose_header_1", "expose_header_2" },
+                @CorsDecorator(origins = "*", exposedHeaders = { "expose_header_1", "expose_header_2" },
                         allowedRequestHeaders = { "allow_request_1", "allow_request_2" },
                         allowedRequestMethods = HttpMethod.GET, maxAge = 3600,
-                        preflightRequestHeaders = {
-                                @AdditionalHeader(name = "x-preflight-cors",
-                                        value = { "Hello CORS", "Hello CORS2" })
-                        })
+                        preflightRequestHeaders = @AdditionalHeader(name = "x-preflight-cors",
+                                value = { "Hello CORS", "Hello CORS2" }))
                 public HttpResponse anyoneGet() {
                     return HttpResponse.of(HttpStatus.OK);
                 }
 
                 @Post("/one/post")
-                @CorsDecorator(
-                        origins = "http://example.com",
+                @CorsDecorator(origins = "http://example.com",
                         exposedHeaders = { "expose_header_1", "expose_header_2" },
                         allowedRequestMethods = HttpMethod.POST, credentialsAllowed = true,
                         allowedRequestHeaders = { "allow_request_1", "allow_request_2" }, maxAge = 1800,
-                        preflightRequestHeaders = {
-                                @AdditionalHeader(name = "x-preflight-cors", value = "Hello CORS")
-                        })
+                        preflightRequestHeaders = @AdditionalHeader(name = "x-preflight-cors",
+                                value = "Hello CORS"))
                 public HttpResponse onePolicyPost() {
                     return HttpResponse.of(HttpStatus.OK);
                 }
 
                 @Get("/multi/get")
-                @CorsDecorator(
-                        origins = "http://example.com", exposedHeaders = { "expose_header_1" },
+                @CorsDecorator(origins = "http://example.com", exposedHeaders = "expose_header_1",
                         allowedRequestMethods = HttpMethod.GET, credentialsAllowed = true)
-                @CorsDecorator(
-                        origins = "http://example2.com", exposedHeaders = { "expose_header_2" },
+                @CorsDecorator(origins = "http://example2.com", exposedHeaders = "expose_header_2",
                         allowedRequestMethods = HttpMethod.GET, credentialsAllowed = true)
                 public HttpResponse multiGet() {
                     return HttpResponse.of(HttpStatus.OK);

--- a/core/src/test/java/com/linecorp/armeria/server/cors/HttpServerCorsTest.java
+++ b/core/src/test/java/com/linecorp/armeria/server/cors/HttpServerCorsTest.java
@@ -57,8 +57,9 @@ public class HttpServerCorsTest {
 
     private static final ClientFactory clientFactory = ClientFactory.DEFAULT;
 
-    @CorsDecorators(value = @CorsDecorator(origins = "http://example.com", exposedHeaders = "expose_header_1"),
-            shortCircuit = true)
+    @CorsDecorators(value = {
+            @CorsDecorator(origins = "http://example.com", exposedHeaders = "expose_header_1")
+    }, shortCircuit = true)
     private static class MyAnnotatedService {
 
         @Get("/index")
@@ -172,30 +173,37 @@ public class HttpServerCorsTest {
             sb.annotatedService("/cors6", new Object() {
 
                 @Get("/any/get")
-                @CorsDecorator(origins = "*", exposedHeaders = { "expose_header_1", "expose_header_2" },
+                @CorsDecorator(
+                        origins = "*", exposedHeaders = { "expose_header_1", "expose_header_2" },
                         allowedRequestHeaders = { "allow_request_1", "allow_request_2" },
                         allowedRequestMethods = HttpMethod.GET, maxAge = 3600,
-                        preflightRequestHeaders = @AdditionalHeader(name = "x-preflight-cors",
-                                value = { "Hello CORS", "Hello CORS2" }))
+                        preflightRequestHeaders = {
+                                @AdditionalHeader(name = "x-preflight-cors",
+                                        value = { "Hello CORS", "Hello CORS2" })
+                        })
                 public HttpResponse anyoneGet() {
                     return HttpResponse.of(HttpStatus.OK);
                 }
 
                 @Post("/one/post")
-                @CorsDecorator(origins = "http://example.com",
+                @CorsDecorator(
+                        origins = "http://example.com",
                         exposedHeaders = { "expose_header_1", "expose_header_2" },
                         allowedRequestMethods = HttpMethod.POST, credentialsAllowed = true,
                         allowedRequestHeaders = { "allow_request_1", "allow_request_2" }, maxAge = 1800,
-                        preflightRequestHeaders = @AdditionalHeader(name = "x-preflight-cors",
-                                value = "Hello CORS"))
+                        preflightRequestHeaders = {
+                                @AdditionalHeader(name = "x-preflight-cors", value = "Hello CORS")
+                        })
                 public HttpResponse onePolicyPost() {
                     return HttpResponse.of(HttpStatus.OK);
                 }
 
                 @Get("/multi/get")
-                @CorsDecorator(origins = "http://example.com", exposedHeaders = "expose_header_1",
+                @CorsDecorator(
+                        origins = "http://example.com", exposedHeaders = { "expose_header_1" },
                         allowedRequestMethods = HttpMethod.GET, credentialsAllowed = true)
-                @CorsDecorator(origins = "http://example2.com", exposedHeaders = "expose_header_2",
+                @CorsDecorator(
+                        origins = "http://example2.com", exposedHeaders = { "expose_header_2" },
                         allowedRequestMethods = HttpMethod.GET, credentialsAllowed = true)
                 public HttpResponse multiGet() {
                     return HttpResponse.of(HttpStatus.OK);

--- a/spring/boot-actuator-autoconfigure/src/test/java/com/linecorp/armeria/spring/actuate/ArmeriaSpringActuatorAutoConfigurationTest.java
+++ b/spring/boot-actuator-autoconfigure/src/test/java/com/linecorp/armeria/spring/actuate/ArmeriaSpringActuatorAutoConfigurationTest.java
@@ -192,9 +192,9 @@ public class ArmeriaSpringActuatorAutoConfigurationTest {
         final String loggerPath = "/internal/actuator/loggers/" + TEST_LOGGER_NAME;
         final AggregatedHttpResponse res =
                 client.execute(RequestHeaders.of(HttpMethod.POST, loggerPath),
-                               OBJECT_MAPPER.writeValueAsBytes(ImmutableMap.of("configuredLevel", "info")))
+                               OBJECT_MAPPER.writeValueAsBytes(ImmutableMap.of("configuredLevel", "debug")))
                       .aggregate().get();
-        assertThat(res.status()).isEqualTo(HttpStatus.UNSUPPORTED_MEDIA_TYPE);
+        assertThat(res.status()).isEqualTo(HttpStatus.NO_CONTENT);
     }
 
     @Test

